### PR TITLE
Refactor search view to make use of permission_classes

### PR DESCRIPTION
**Pull request by stsewd**
This is the same as https://github.com/readthedocs/readthedocs.org/pull/6133
but for the search view.

This allows us to override `get_all_projects` in .com
to filter projects using the auth backends.

And make use of the permission class from .com that makes use of the
auth backends.

This doesn't break the current search endpoint in .com
since it filters the versions using `public(user=self.request.user)`

A change is that now when a project or version doesn't exist, a 404 is raised, and we allow search on PR versions.

This is required to proxy search on .com

### CODE REVIEW

File: readthedocs/search/api.py [hunk 1]
```diff
@@ -78,13 +119,7 @@ def get_queryset(self):
         query = self.request.query_params.get('q', '')
         kwargs = {'filter_by_user': False, 'filters': {}}
         kwargs['filters']['project'] = [p.slug for p in self.get_all_projects()]
-        kwargs['filters']['version'] = self.request.query_params.get('version')
-        if not kwargs['filters']['project']:
-            log.info("Unable to find a project to search")
-            return HTMLFile.objects.none()
-        if not kwargs['filters']['version']:
-            log.info("Unable to find a version to search")
-            return HTMLFile.objects.none()
```
Comment (ericholscher) [inline readthedocs/search/api.py]: Why did we remove this validation? It is important that we keep it to avoid another security issue like the last time we messed this up.

File: readthedocs/search/utils.py [hunk 2]
```diff
@@ -82,27 +82,6 @@ def remove_indexed_files(model, project_slug, version_slug=None, build_id=None):
...
 
 
-# TODO: Rewrite all the views using this in Class Based View,
-# and move this function to a mixin
-def get_project_list_or_404(project_slug, user, version_slug=None):
-    """
-    Return list of project and its subprojects.
-
-    It filters by Version privacy instead of Project privacy,
-    so we can support public versions on private projects.
```
Comment (ericholscher) [inline readthedocs/search/utils.py]: We lost this comment. Why did we remove this function in this PR? I'd really like to keep these changes small, and this PR has a number of different refactors all put together, which makes it harder to think about.

File: readthedocs/search/api.py [hunk 3]
```diff
@@ -120,17 +155,24 @@ def get_all_projects(self):
...
+        main_version = self._get_version()
+        main_project = self._get_project()
+
+        subprojects = Project.objects.filter(
+            superprojects__parent_id=main_project.id,
         )
+        all_projects = []
+        for project in list(subprojects) + [main_project]:
+            version = (
+                Version.objects
```
Comment (ericholscher) [inline readthedocs/search/api.py]: This is using `objects` instead of `internal`.

File: readthedocs/search/api.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/search/api.py]: This block happens when project or version are None, if project or version are None, a 404 is raised in the code above. So this block isn't needed.

Comment (ericholscher) [reply readthedocs/search/api.py]: What happens if `get_all_projects` returns an empty list?

File: readthedocs/search/utils.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/search/utils.py]: It wasn't used in any other part, I can put back the comment about privacy in `get_all_projects`

File: readthedocs/search/api.py (hunk 1 above)
Comment (ericholscher) [reply readthedocs/search/api.py]: Code doesn't just get added to our repos for no reason. You really need to be careful when deleting or changing things, and understand _why_ they were there. This would likely have caused another major security issue with search if it was shipped.

Comment (stsewd) [reply readthedocs/search/api.py]: I can't think of a case where it returns empty, but I'm adding a test mocking it to return empty

Comment (ericholscher) [reply readthedocs/search/api.py]: Where a project doesn't have any public versions and the user doesn't have access to a private version. It happened in the past, and maybe the code is more defensive now, but I'd really like to keep this additional check just to be sure.

Comment (stsewd) [reply readthedocs/search/api.py]: Check is back with a test!

File: readthedocs/search/api.py [hunk 4]
```diff
@@ -118,19 +157,26 @@ def get_serializer_context(self):
...
+        for project in list(subprojects) + [main_project]:
+            version = (
+                Version.internal
+                .public(user=self.request.user, project=project)
+                .filter(slug=main_version.slug)
+                .first()
+            )
+            if version:
+                all_projects.append(version.project)
         return all_projects
```
Comment (ericholscher) [inline readthedocs/search/api.py]: We should probably rename this `all_versions`, since it's actually a list of versions, not projects :)

File: readthedocs/search/api.py [hunk 5]
```diff
@@ -60,11 +62,50 @@ def get_inner_hits(self, obj):
...
+        if not version:
+            version_slug = self.request.GET.get('version', None)
+            project = self._get_project()
+            version = get_object_or_404(
+                project.versions.all(),
+                slug=version_slug,
+            )
+            setattr(self, cache_key, version)
+
+        return version
```
Comment (ericholscher) [inline readthedocs/search/api.py]: Why do we need caching here?

File: readthedocs/search/api.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/search/api.py]: It's a list of projects, version is used to validate the user has access to the version of that subproject before searching.

File: readthedocs/search/api.py (hunk 5 above)
Comment (stsewd) [reply readthedocs/search/api.py]: It's called 3 times, so we don't do the same query 3 times, there is a comment on the docstring from the class

File: readthedocs/search/api.py [hunk 6]
```diff
@@ -78,13 +119,11 @@ def get_queryset(self):
...
         kwargs['filters']['project'] = [p.slug for p in self.get_all_projects()]
-        kwargs['filters']['version'] = self.request.query_params.get('version')
+        kwargs['filters']['version'] = self._get_version().slug
+        # Check to avoid searching all projects in case project is empty.
         if not kwargs['filters']['project']:
             log.info("Unable to find a project to search")
             return HTMLFile.objects.none()
-        if not kwargs['filters']['version']:
-            log.info("Unable to find a version to search")
-            return HTMLFile.objects.none()
```
Comment (ericholscher) [inline readthedocs/search/api.py]: I'd still like to keep this as well, just to be safe.
